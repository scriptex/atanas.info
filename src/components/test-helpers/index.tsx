import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '../../scripts/shared';

export const mockUseEffect = (): void => {
	let useEffect: jest.Mock<React.EffectCallback>;

	const mockedUseEffect = () => {
		useEffect.mockImplementationOnce((f: () => React.EffectCallback) => f());
	};

	beforeEach(() => {
		useEffect = jest.spyOn(React, 'useEffect') as unknown as jest.Mock<React.EffectCallback>;

		mockedUseEffect();
	});
};

export const snapshotTest = (
	Component: React.FC<Readonly<React.ComponentProps<any>>>,
	buttonSelector?: string
): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	mockUseEffect();

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { asFragment, container } = result;

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				act(() => {
					container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { unmount, asFragment } = result;

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};
