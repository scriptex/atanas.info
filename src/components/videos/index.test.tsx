import { Videos } from '.';
import { snapshotTest } from '../test-helpers';

snapshotTest(Videos);
