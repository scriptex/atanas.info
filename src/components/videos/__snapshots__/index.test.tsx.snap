// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Videos component Should render the Videos component 1`] = `
<DocumentFragment>
  <main
    class="o-main"
  >
    <section
      class="c-section c-section--videos c-section--slides"
      id="videos"
    >
      <div
        class="o-shell"
      >
        <header
          class="c-section__header"
        >
          <div
            class="c-section__animation"
            style="width: 150px; height: 150px; margin-left: auto; margin-right: auto;"
          />
          <h2>
            Videos
          </h2>
        </header>
        <div
          class="lines"
        >
          <div
            class="line line--vertical"
          />
          <div
            class="line line--vertical"
          />
          <div
            class="line line--vertical"
          />
        </div>
        <div
          class="lines"
        >
          <div
            class="line line--horizontal"
          />
          <div
            class="line line--horizontal"
          />
          <div
            class="line line--horizontal"
          />
        </div>
        <nav
          class="c-section__nav"
        >
          <ul>
            <li
              class="current"
            >
              <button
                aria-label="Switch to Do more with Sublime Text"
                class="c-btn"
                type="button"
              >
                Do more with Sublime Text
              </button>
            </li>
            <li>
              <button
                aria-label="Switch to Jarvis: Frontend project setup"
                class="c-btn"
                type="button"
              >
                Jarvis: Frontend project setup
              </button>
            </li>
          </ul>
        </nav>
        <div
          class="c-section__body"
        >
          <div
            class="c-section__frame current"
          >
            <div
              class="c-loading"
            >
              <div
                class="c-loader"
              />
            </div>
            <lite-youtube
              params="autoplay=0"
              playlabel="Do more with Sublime Text"
              style="background-image: url(https://i.ytimg.com/vi/5xcqprI17FQ/hqdefault.jpg);"
              title="Presentation about optimizing the web development workflow using Sublime Text editor"
              videoid="5xcqprI17FQ"
            >
              <button
                class="lty-playbtn"
                type="button"
              >
                <span
                  class="lyt-visually-hidden"
                >
                  Do more with Sublime Text
                </span>
              </button>
            </lite-youtube>
          </div>
          <div
            class="c-section__frame"
          >
            <div
              class="c-loading"
            >
              <div
                class="c-loader"
              />
            </div>
            <lite-youtube
              params="autoplay=0"
              playlabel="Jarvis: Frontend project setup"
              style="background-image: url(https://i.ytimg.com/vi/s-klXHLIZyY/hqdefault.jpg);"
              title="Presentation about the new and optimized frontend projects setup"
              videoid="s-klXHLIZyY"
            >
              <button
                class="lty-playbtn"
                type="button"
              >
                <span
                  class="lyt-visually-hidden"
                >
                  Jarvis: Frontend project setup
                </span>
              </button>
            </lite-youtube>
          </div>
        </div>
        <div
          class="c-section__actions"
        >
          <button
            class="c-btn"
            type="button"
          >
            Get in touch
          </button>
          <ul
            class="c-section__actions-list"
          >
            <li>
              <button
                class="c-btn c-section__actions-close"
                type="button"
              >
                Close
              </button>
            </li>
            <li>
              <button
                class="c-btn"
                type="button"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-email"
                  />
                </svg>
                <span>
                  Contact me
                </span>
                <strong>
                  directly
                </strong>
              </button>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://app.usebraintrust.com/talent/782/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-braintrust"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Braintrust
                </strong>
              </a>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://www.toptal.com/resume/atanas-atanasov"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-toptal"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Toptal
                </strong>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`Videos component Should unmount the Videos component 1`] = `<DocumentFragment />`;
